File: java/Dson-Core/src/main/java/cn/wjybxx/dson/DsonFv4.java
/*
 * Copyright 2026 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson;

import cn.wjybxx.dson.types.FixedVector4;

import javax.annotation.Nonnull;
import java.util.Objects;

/** @author wjybxx */
public final class DsonFv4 extends DsonValue {

    public static final DsonFv4 EMPTY = new DsonFv4(FixedVector4.EMPTY);

    private final FixedVector4 value;

    public DsonFv4(FixedVector4 value) {
        this.value = Objects.requireNonNull(value);
    }

    public FixedVector4 getValue() {
        return value;
    }

    @Nonnull
    @Override
    public DsonType getDsonType() {
        return DsonType.FIXED_VECTOR4;
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        DsonFv4 that = (DsonFv4) o;
        return value.equals(that.value);
    }

    @Override
    public int hashCode() {
        return value.hashCode();
    }

    @Override
    public String toString() {
        return "DsonFv4{" +
                "value=" + value +
                '}';
    }
}

File: csharp/Wjybxx.Dson.Core/src/DsonFv4.cs
using System;
using Wjybxx.Dson.Types;

namespace Wjybxx.Dson
{
public sealed class DsonFv4 : DsonValue, IEquatable<DsonFv4>
{
    public static readonly DsonFv4 EMPTY = new DsonFv4(default);
    private readonly FixedVector4 _value;

    public DsonFv4(FixedVector4 value) {
        _value = value;
    }

    public override DsonType DsonType => DsonType.FixedVector4;
    public FixedVector4 Value => _value;

    public bool Equals(DsonFv4? other) {
        if (ReferenceEquals(null, other)) return false;
        if (ReferenceEquals(this, other)) return true;
        return _value.Equals(other._value);
    }

    public override bool Equals(object? obj) {
        return obj is DsonFv4 other && Equals(other);
    }

    public override int GetHashCode() {
        return _value.GetHashCode();
    }

    public static bool operator ==(DsonFv4? left, DsonFv4? right) {
        return Equals(left, right);
    }

    public static bool operator !=(DsonFv4? left, DsonFv4? right) {
        return !Equals(left, right);
    }

    public override string ToString() {
        return $"{nameof(DsonType)}: {DsonType}, {nameof(_value)}: {_value}";
    }
}
}